src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java

```diff
@@ -65,20 +64,16 @@ public EntitiesPopulator(CaveBlock addon)
     @Override
     public void populate(World world, Random random, Chunk chunk)
     {
-        for (Map.Entry<EntityType, Pair<Integer, Integer>> entry : chances.get(world.getEnvironment()).entityChanceMap.entrySet())
+        for (Map.Entry<EntityType, Pair<Double, Integer>> entry : chances.get(world.getEnvironment()).entityChanceMap.entrySet())
         {
             for (int subY = 0; subY < worldHeight; subY += 16)
             {
-                for (int tries = 0; tries < generationTry; tries++)
+                // Use double so chance can be < 1
+                if (random.nextDouble() * 100 < entry.getValue().x)
                 {
-                    if (random.nextInt(100) < entry.getValue().x)
-                    {
-                        int x = random.nextInt(15);
-                        int z = random.nextInt(15);
-                        int y = Math.min(worldHeight - 2, subY + random.nextInt(15));
-
-                        this.tryToPlaceEntity(world, chunk.getBlock(x, y, z), entry.getKey(), x, z, chances.get(world.getEnvironment()).mainMaterial);
-                    }
+                    int y = Math.min(worldHeight - 2, subY + random.nextInt(15));
+                    // Spawn only in middle of chunk because bounding box will grow out from here
+                    this.tryToPlaceEntity(world, chunk.getBlock(7, y, 7), entry.getKey(), chances.get(world.getEnvironment()).mainMaterial);
```

Comment:
This is a simplification so it means the positioning is less random, but I don't think it'll matter too much.